packages/core/test/lib/utils/promisebuffer.test.ts

```diff
@@ -1,52 +1,163 @@
+
+      void buffer.add(producer1);
+      void buffer.add(producer2);
+      void buffer.add(producer3);
+      void buffer.add(producer4);
+      void buffer.add(producer5);
+      await expect(buffer.add(producer6)).rejects.toThrowError();
+
+      expect(producer1).toHaveBeenCalledTimes(1);
+      expect(producer2).toHaveBeenCalledTimes(1);
+      expect(producer3).toHaveBeenCalledTimes(1);
+      expect(producer4).toHaveBeenCalledTimes(1);
+      expect(producer5).toHaveBeenCalledTimes(1);
+      expect(producer6).not.toHaveBeenCalled();
+
+      expect(buffer.$.length).toEqual(5);
+
+      await buffer.drain();
+
+      expect(buffer.$.length).toEqual(0);
+
+      expect(producer1).toHaveBeenCalledTimes(1);
+      expect(producer2).toHaveBeenCalledTimes(1);
+      expect(producer3).toHaveBeenCalledTimes(1);
+      expect(producer4).toHaveBeenCalledTimes(1);
+      expect(producer5).toHaveBeenCalledTimes(1);
+      expect(producer6).not.toHaveBeenCalled();
+    });
+
+    test('sync promises', async () => {
+      const buffer = makePromiseBuffer(1);
+      let task1;
+      const producer1 = vi.fn(() => {
+        task1 = resolvedSyncPromise();
+        return task1;
+      });
+      const producer2 = vi.fn(() => resolvedSyncPromise());
+      expect(buffer.add(producer1)).toEqual(task1);
+      const add2 = buffer.add(producer2);
+
+      // This is immediately executed and removed again from the buffer
+      expect(buffer.$.length).toEqual(0);
+
+      await expect(add2).resolves.toBeUndefined();
+
+      expect(producer1).toHaveBeenCalled();
+      expect(producer2).toHaveBeenCalled();
     });
 
-    test('with limit', () => {
+    test('async promises', async () => {
       const buffer = makePromiseBuffer(1);
       let task1;
       const producer1 = vi.fn(() => {
-        task1 = new SyncPromise(resolve => setTimeout(resolve));
+        task1 = new Promise(resolve => setTimeout(resolve, 1));
         return task1;
       });
-      const producer2 = vi.fn(() => new SyncPromise(resolve => setTimeout(resolve)));
+      const producer2 = vi.fn(() => new Promise(resolve => setTimeout(resolve, 1)));
       expect(buffer.add(producer1)).toEqual(task1);
-      void expect(buffer.add(producer2)).rejects.toThrowError();
+      const add2 = buffer.add(producer2);
+
       expect(buffer.$.length).toEqual(1);
+
+      await expect(add2).rejects.toThrowError();
+
       expect(producer1).toHaveBeenCalled();
       expect(producer2).not.toHaveBeenCalled();
     });
+
+    test('handles multiple equivalent promises', async () => {
+      const buffer = makePromiseBuffer(10);
+
+      const promise = new Promise(resolve => setTimeout(resolve, 1));
+
+      const producer = vi.fn(() => promise);
+      const producer2 = vi.fn(() => promise);
+
+      expect(buffer.add(producer)).toEqual(promise);
+      expect(buffer.add(producer2)).toEqual(promise);
+
+      expect(buffer.$.length).toEqual(1);
```

Comment:
Do you think this is a case we could even run into in real life? (don't think this is wrong per sé just a bit interesting because we handle this differently e.g. in client hook subscribers)